src/arrays.jl

```diff
@@ -727,7 +727,12 @@ function scalarize(arr)
         end
     elseif istree(arr) && operation(arr) == getindex
         args = arguments(arr)
-        scalarize(args[1], (args[2:end]...,))
+        scalarized = scalarize(args[1], (args[2:end]...,))
+        if isnothing(metadata(arr))
+            scalarized
+        else
+            metadata(scalarized, metadata(arr))
```

Comment:
@shashi @YingboMa should the metadata be the same, or should it know that it's derived from indexing this object somehow?

Reply:
It should not be the same... And I think this should not be added... I would say this is a usecase where tagged-data like API becomes useful. You should be able to define something like:

```julia

@metadata getindex(::MetadataType, idx) = compute_output_metadata_here
```

And it should be handled for every key which registers in this way.

I'd be interested in adding this general feature, but I'm looking for more substantial examples where this is super useful.

Until then, I think you should just try to propagate the metadata outside of the package's behavior (by walking the tree once constructed). The method in this PR seems not correct to me.